crates/bitwarden-vault/src/tree.rs

```diff
@@ -0,0 +1,275 @@
+use std::collections::HashMap;
+use std::fmt::Debug;
+use uuid::Uuid;
+
+pub trait TreeItem: Clone + Debug {
+    fn id(&self) -> Option<Uuid>;
+    /*
+    This is the name that will be output when getting the tree nodes
+     */
+    fn short_name(&self) -> &str;
+    /*
+    This is the path that the item is stored into a tree
+     */
+    fn path(&self) -> Vec<&str>;
+    const DELIMITER: char;
+}
+
+#[derive(Clone, Debug)]
+pub struct TreeIndex<T: TreeItem> {
+    pub id: usize, // location in the tree
+    pub data: T, // this will be the raw value
+    pub path: Vec<String>
+}
+
+impl<T: TreeItem> TreeIndex<T> {
+    pub fn new(id: usize, data: T) -> Self {
+        TreeIndex {
+            id,
+            data: data.clone(),
+            path: data.path().iter().map(|s| s.to_string()).collect(),
+        }
+    }
+}
+
+pub struct NodeItem<T: TreeItem> {
+    pub item: T,
+    pub parent: Option<T>,
+    pub children: Vec<T>
+}
+
+pub struct TreeNode {
+    pub id: usize,
+    pub item_id: Uuid,
+    pub parent_idx: Option<usize>,
+    pub children_idx: Vec<usize>,
+    pub path: Vec<String>
+}
+
+impl TreeNode {
+    pub fn new<T: TreeItem>(id: usize, parent_idx: Option<usize>, children_idx: Vec<usize>, index: TreeIndex<T>) -> Self {
+        TreeNode {
+            id,
+            item_id: index.data.id().unwrap(),
+            parent_idx,
+            children_idx,
+            path: index.path,
+        }
+    }
+}
+
+pub struct Tree<T: TreeItem> {
+    pub nodes: Vec<TreeNode>,
+    pub items: Vec<TreeIndex<T>>,
+    path_to_node: HashMap<Vec<String>, usize>
+}
+
+impl<T: TreeItem> Tree<T> {
+    pub fn from_items(items: &mut Vec<T>) -> Self {
```

Comment:
suggestion: I would take the value here since we are modifying it in unexpected ways inside of the function (sorting). It might also mean that we can get away with not cloning the items when you put them in the tree since you'd have ownership of them